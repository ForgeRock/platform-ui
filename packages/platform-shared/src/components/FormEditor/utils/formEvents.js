/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { cloneDeep } from 'lodash';

/**
 * Executes a given script in a Web Worker, providing form values and schema as variables.
 * The worker exists only to run the script and return the potentially modified form values and schema.
 * Worker is terminated after recieving the result.
 *
 * @param {string} script - The JavaScript code to execute inside the Web Worker.
 * @param {Object} formValues - The current form values to be passed to the worker.
 * @param {Object} formSchema - The form schema to be passed to the worker.
 * @returns {Promise<{formValues: Object, formSchema: Object}>} Resolves with the potentially updated form values and schema.
 * @throws {Error} If the worker encounters an error during script execution.
 */
// eslint-disable-next-line import/prefer-default-export
export async function useWebWorker(script, formValues, formSchema) {
  return new Promise((resolve, reject) => {
    const worker = new Worker(new URL('./formEventsWorker.js', import.meta.url));

    worker.onmessage = (e) => {
      if (e.data.error) {
        reject(new Error(`Error executing onLoad event: ${e.data.error}`));
        worker.terminate();
        return;
      }

      resolve({
        formValues: e.data.formValues,
        formSchema: e.data.formSchema,
      });

      worker.terminate();
    };

    const scriptVariables = {
      formValues: cloneDeep(formValues),
      formSchema: cloneDeep(formSchema),
    };

    worker.postMessage({
      script, scriptVariables,
    });
  });
}
