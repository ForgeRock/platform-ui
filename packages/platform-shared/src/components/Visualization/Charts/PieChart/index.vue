<!-- Copyright (c) 2022-2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div class="position-relative">
    <div
      data-testid="chart"
      ref="d3chart"
      :id="id" />
    <div
      data-testid="legend"
      :class="[{ 'd-flex justify-content-center mt-4': !legendClass.length}, legendClass]">
      <ul
        v-if="legend.length"
        class="list-unstyled">
        <li
          v-for="item in legend"
          :key="item.label">
          <span class="d-flex align-items-center">
            <div
              class="rounded-pill mr-3"
              :style="{ height: '10px', width: '24px', 'background-color': item.color }" />
            <template v-if="showLegendCount">
              {{ item.label }} ({{ item.value }})
            </template>
            <template v-else>
              {{ item.label }}
            </template>
          </span>
        </li>
      </ul>
    </div>
  </div>
</template>

<script>
import * as d3 from 'd3';

/*
 * @description Pie Chart Component used to show data in a pie chart with percentages according the value of each item,
 *
 * @param {Array}   data    data to be loaded in the pie chart, each item of the array has the following structure
 *                          @property {string}  label - label to be displayed on the items list
 *                          @property {number}  value - value to calculate the percentage used on the pie chart, the
 *                                                      D3PieChart component obtains this percentage by adding up the
 *                                                      value of all the items and based on that total it obtains the
 *                                                      percentage of the value of each item with a rule of three, for
 *                                                      example for the values 20, 50, 70 the total is 140 therefore
 *                                                      for the first item the operation (20 * 100) / 140 is applied
 *                                                      which result is the percentage 14%.
 *                          @property {string}  color - color for the section in the pie chart and the bullet on items list
 *                          {
 *                            label: 'Application',
 *                            value: 30,
 *                            color: '#109cf1'
 *                          }
 * @param {String}  height  height in pixels for the svg where the chart is displayed
 */
export default {
  name: 'PieChart',
  props: {
    data: {
      type: Array,
      default: () => [],
    },
    height: {
      type: Number,
      default: 250,
    },
    hideTooltip: {
      type: Boolean,
      deafault: false,
    },
    id: {
      type: String,
      default: 'PieChart',
    },
    legendClass: {
      type: String,
      default: '',
    },
    radius: {
      type: Number,
      default: 110,
    },
    showLegendCount: {
      type: Boolean,
      default: false,
    },
    strokeWidth: {
      type: Number,
      default: 3,
    },
    width: {
      type: Number,
      default: 250,
    },
  },
  data() {
    return {
      legend: [],
      tooltip: null,
    };
  },
  mounted() {
    this.loadData();
  },
  methods: {
    loadData() {
      const chartData = {};
      const colors = [];
      this.legend = [];

      if (this.data.length) {
        this.data.forEach((category, index) => {
          chartData[index] = category.value;
          colors.push(category.color);
          this.legend.push({
            label: category.label,
            color: category.color,
            value: category.value,
          });
        });

        this.createChart(chartData, colors);
      }
    },
    createChart(chartData, colors) {
      // set the dimensions and margins of the graph
      const margin = 0;

      // The radius of the pieplot is half the width or half the height (smallest one). I subtract a bit of margin.
      const radius = Math.min(this.width, this.height) / 2 - margin;

      d3.select(`#${this.id}`)
        .selectAll('svg').remove();

      // append the svg object to the div called 'my_dataviz'
      const svg = d3
        .select(`#${this.id}`)
        .append('svg')
        .attr('width', this.width)
        .attr('height', this.height)
        .attr('class', 'mx-auto d-block')
        .append('g')
        .attr(
          'transform',
          `translate(${this.width / 2},${this.height / 2})`,
        );

      // Create dummy data
      const data = chartData;

      // set the color scale
      const color = d3
        .scaleOrdinal()
        .domain(Object.keys(data))
        .range(colors);

      // Compute the position of each group on the pie:
      const pie = d3.pie().value((d) => d[1]);

      // eslint-disable-next-line camelcase
      const data_ready = pie(Object.entries(data));

      if (!this.hideTooltip) {
        this.tooltip = d3.select(`#${this.id}`)
          .append('div')
          .style('position', 'fixed')
          .style('visibility', 'hidden')
          .style('background-color', 'transparent')
          .style('border-width', '0px')
          .style('border-radius', '5px')
          .style('padding', '0 0 5px 0');
      }

      // Build the pie chart: Basically, each part of the pie is a path that we build using the arc function.
      svg
        .selectAll('whatever')
        .data(data_ready)
        .enter()
        .append('path')
        .attr('class', `${this.id}-tooltip`)
        .attr(
          'd',
          d3
            .arc()
            .innerRadius(this.radius) // This is the size of the donut hole
            .outerRadius(radius)
            .cornerRadius(10),
        )
        .attr('fill', (d) => color(d.data[0]))
        .attr('stroke', 'white')
        .style('stroke-width', `${this.strokeWidth}px`);

      if (!this.hideTooltip) {
        d3.selectAll(`.${this.id}-tooltip`)
          .on('mouseover', () => this.tooltip.style('visibility', 'visible'))
          .on('mouseout', () => this.tooltip.style('visibility', 'hidden'))
          .on('mousemove', this.mouseMove);
      }
    },
    mouseMove(event) {
      const { width, height } = this.tooltip.node().getBoundingClientRect();
      return this.tooltip
        .style('top', `${event.clientY - (height + 5)}px`)
        .style('left', `${event.clientX - (width / 2)}px`)
        .html(() => {
          const { index } = event.srcElement.__data__;
          return `<div role="tooltip"">
            <div class="arrow"></div>
            <div class="tooltip-inner">
              <div class="text-gray-400">${this.legend[index].label}</div>
              <div class="font-weight-bold">${this.legend[index].value}</div>
              </div>
              </div>
          `;
        });
    },
  },
  watch: {
    data() {
      this.loadData();
    },
  },
};
</script>

<style lang="scss" scoped>
::v-deep {
  .arrow {
    position: absolute;
    display: block;
    width: 100%;
    height: 0.4rem;
    text-align: center;
    bottom: 0;
    left: -5px;

    &::before {
      top: 0;
      border-width: 0.4rem 0.4rem 0;
      border-color: transparent;
      border-top-color: black;
      position: absolute;
      content: "";
      border-style: solid;
    }
  }
}
</style>
