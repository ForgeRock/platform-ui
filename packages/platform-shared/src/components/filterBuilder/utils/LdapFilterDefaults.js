/**
 * Copyright (c) 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

export const ldapOperatorOptions = {
  All: { label: 'All', delimeter: 'and' },
  Any: { label: 'Any', delimeter: 'or' },
};

export const ldapDefaultConditionOptions = {
  Equals: { label: 'is', value: '=' },
  NotEquals: { label: 'is not', value: '!=' },
  Gte: { label: 'GTE (>=)', value: '>=' },
  Gt: { label: 'GT (>)', value: '!<=' },
  Lte: { label: 'LTE (<=)', value: '<=' },
  Lt: { label: 'LT (<)', value: '!>=' },
  BitwiseAnd: { label: 'Bitwise AND', value: ':1.2.840.113556.1.4.803:=' },
  BitwiseNand: { label: 'Bitwise NAND', value: '!:1.2.840.113556.1.4.803:=' },
  BitwiseOr: { label: 'Bitwise OR', value: ':1.2.840.113556.1.4.804:=' },
  BitwiseNor: { label: 'Bitwise NOR', value: '!:1.2.840.113556.1.4.804:=' },
};

export const conditionMap = {
  equal: '=',
  ge: '>=',
  le: '<=',
};

export function getTypeFromValue(value, properties) {
  if (!value) return 'string';
  const property = properties.find((prop) => prop.value === value);
  return property ? property.type : 'string';
}
