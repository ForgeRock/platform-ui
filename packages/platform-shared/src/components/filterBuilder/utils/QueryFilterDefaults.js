/**
 * Copyright (c) 2020-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import i18n from '@/i18n';

export const operatorOptions = {
  All: { label: i18n.global.t('filters.all'), delimeter: 'and' },
  Any: { label: i18n.global.t('filters.any'), delimeter: 'or' },
};

export const defaultConditionOptions = {
  Contains: { label: i18n.global.t('filters.contains'), value: 'co', type: ['string'] },
  DoesNotContain: { label: i18n.global.t('filters.doesNotContain'), value: '!co', type: ['string'] },
  Is: { label: i18n.global.t('filters.is'), value: 'eq', type: ['string', 'number', 'boolean'] },
  IsNot: { label: i18n.global.t('filters.isNot'), value: '!eq', type: ['string', 'number'] },
  IsPresent: { label: i18n.global.t('filters.isPresent'), value: 'pr', type: ['string', 'number', 'boolean'] },
  IsNotPresent: { label: i18n.global.t('filters.isNotPresent'), value: '!pr', type: ['string', 'number', 'boolean'] },
  StartsWith: { label: i18n.global.t('filters.startsWith'), value: 'sw', type: ['string'] },
  DoesNotStartsWith: { label: i18n.global.t('filters.doesNotStartWith'), value: '!sw', type: ['string'] },
  Gte: { label: i18n.global.t('filters.greaterThanEqual'), value: 'ge', type: ['number'] },
  Gt: { label: i18n.global.t('filters.greaterThan'), value: 'gt', type: ['number'] },
  Lte: { label: i18n.global.t('filters.lessThanEqual'), value: 'le', type: ['number'] },
  Lt: { label: i18n.global.t('filters.lessThan'), value: 'lt', type: ['number'] },
};

export const governanceConditionOptions = {
  Contains: { label: i18n.global.t('filters.contains'), value: 'contains', type: ['string', 'array'] },
  DoesNotContain: { label: i18n.global.t('filters.doesNotContain'), value: 'not_contains', type: ['string', 'array'] },
  Is: { label: i18n.global.t('filters.is'), value: 'equals', type: ['string', 'reference', 'boolean', 'number'] },
  IsNot: { label: i18n.global.t('filters.isNot'), value: 'not_equals', type: ['string', 'reference', 'number'] },
  StartsWith: { label: i18n.global.t('filters.startsWith'), value: 'starts_with', type: ['string'] },
  DoesNotStartsWith: { label: i18n.global.t('filters.endsWith'), value: 'ends_with', type: ['string'] },
  HasChanged: { label: i18n.global.t('filters.hasChanged'), value: 'has_changed', type: ['reference', 'string', 'number'] },
  HasNotChanged: { label: i18n.global.t('filters.hasNotChanged'), value: 'has_not_changed', type: ['reference', 'string', 'number'] },
  Gte: { label: i18n.global.t('filters.greaterThanEqual'), value: 'gte', type: ['number'] },
  Gt: { label: i18n.global.t('filters.greaterThan'), value: 'gt', type: ['number'] },
  Lte: { label: i18n.global.t('filters.lessThanEqual'), value: 'lte', type: ['number'] },
  Lt: { label: i18n.global.t('filters.lessThan'), value: 'lt', type: ['number'] },
};

export const temporalValueOptions = [
  { text: i18n.global.t('filters.previousValueOf'), value: 'before' },
  { text: i18n.global.t('filters.currentValueOf'), value: 'after' },
];

export function getTypeFromValue(value, properties) {
  if (!value) return 'string';
  const property = properties.find((prop) => prop.value === value);
  return property ? property.type : 'string';
}
