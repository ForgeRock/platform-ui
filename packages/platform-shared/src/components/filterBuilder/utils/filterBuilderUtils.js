/**
 * Copyright (c) 2022-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

const filterFieldMap = {
  starts_with: 'value',
  not_starts_with: 'value',
  ends_with: 'value',
  has_changed: 'left',
  has_not_changed: 'left',
  not_ends_with: 'value',
  equals: 'left',
  not_equals: 'left',
  gte: 'left',
  gt: 'left',
  lte: 'left',
  lt: 'left',
  contains: 'in_string',
  not_contains: 'in_string',
};
const filterValueMap = {
  starts_with: 'prefix',
  not_starts_with: 'prefix',
  ends_with: 'suffix',
  not_ends_with: 'suffix',
  equals: 'right',
  not_equals: 'right',
  gte: 'right',
  gt: 'right',
  lte: 'right',
  lt: 'right',
  contains: 'search_string',
  not_contains: 'search_string',
};
const operatorMap = {
  has_changed: 'not_equals',
  has_not_changed: 'equals',
};

/**
 * Converts filter from shared filter builder format to IGA format
 * {
 *   operator: 'or',
 *   subfilters: [
 *     {
 *       field: 'sn',
 *       operator: 'starts_with',
 *       value: 'koya',
 *       temporalValue: 'before',
 *     },
 *   ],
 * },
 * {
 *   or: [
 *     {
 *       starts_with: {
 *         prefix: {
 *           literal: 'koya',
 *         },
 *         value: 'user.before.sn',
 *       },
 *     },
 *   ],
 * },
 * @param {Object} filter component filter to convert
 * @param {String} resourceName name of resource filter is aimed against
 * @param {Array} properties list of properties available
 */
export function convertToIGAFilter(filter, resourceName = 'user', properties, ignoreTemporals) {
  const convertedFilter = {
    [filter.operator]: [],
  };
  filter.subfilters.forEach((filterRule) => {
    if (filterRule.operator === 'or' || filterRule.operator === 'and') {
      convertedFilter[filter.operator].push(convertToIGAFilter(filterRule, resourceName, properties, ignoreTemporals));
    } else {
      const operator = operatorMap[filterRule.operator] || filterRule.operator;
      const subfilter = { [operator]: {} };
      if (filterRule.operator.endsWith('_changed')) {
        // Special consideration for has changed and has not changed
        subfilter[operator].left = `${resourceName}.before.${filterRule.field}`;
        subfilter[operator].right = `${resourceName}.after.${filterRule.field}`;
      } else {
        subfilter[operator][filterValueMap[operator]] = typeof filterRule.value === 'string' && filterRule.value.startsWith(`${resourceName}.`)
          ? filterRule.value
          : { literal: filterRule.value };
        const selectedProperty = properties.find((property) => property.value === filterRule.field);

        // convert boolean values from string to boolean
        if (selectedProperty?.type === 'boolean') {
          subfilter[operator][filterValueMap[operator]] = { literal: filterRule.value === 'True' };
        }

        if (selectedProperty?.type === 'array' && selectedProperty.path) {
          // We need to set arrays of relationships back to in_string_array instead of in_string
          subfilter[operator].in_string_array = ignoreTemporals
            ? filterRule.field
            : `${resourceName}.${filterRule.temporalValue}.${filterRule.field}`;
        } else {
          subfilter[operator][filterFieldMap[operator]] = ignoreTemporals
            ? filterRule.field
            : `${resourceName}.${filterRule.temporalValue}.${filterRule.field}`;
        }
      }
      convertedFilter[filter.operator].push(subfilter);
    }
  });
  return convertedFilter;
}

/**
 * Takes IGA formatted filter and converts to format that can be consumed by shared filter builder
 * @property {Object} filter - IGA filter to convert
 * @returns {Object} filterBuilder-formatted filter
 */
export function convertFromIGAFilter(filter, currentUniqueIndex = 0, ignoreTemporals) {
  let uniqueIndex = currentUniqueIndex;
  const igaFilter = Object.keys(filter).length ? filter : { or: [{ starts_with: { prefix: { literal: '' }, value: 'user.after.' } }] };
  const filterEntry = Object.entries(igaFilter)[0];
  const convertedFilter = {
    operator: filterEntry[0],
    subfilters: [],
    uniqueIndex,
  };
  filterEntry[1].forEach((filterRule) => {
    const [operator, fieldValueObject] = Object.entries(filterRule)[0];
    if (operator === 'or' || operator === 'and') {
      const subfilter = convertFromIGAFilter(filterRule, uniqueIndex + 1, ignoreTemporals);
      convertedFilter.subfilters.push(subfilter.convertedFilter);
      uniqueIndex = subfilter.uniqueIndex;
    } else {
      uniqueIndex += 1;
      const subfilter = {
        operator,
        uniqueIndex,
      };
      // equals and not_equals are used fot the has_changed and has_not_changed operators
      // these operators will have left and right string values as opposed to a regular
      // equals or not_equals which will have an object as the right value
      if ((operator === 'not_equals' || operator === 'equals')
        && (typeof fieldValueObject.left === 'string' || fieldValueObject.left instanceof String)
        && (typeof fieldValueObject.right === 'string' || fieldValueObject.right instanceof String)) {
        // convert to has_changed and has_not_changed
        const field = fieldValueObject.left?.split('.').pop();
        subfilter.operator = operator === 'not_equals'
          ? 'has_changed'
          : 'has_not_changed';
        subfilter.field = field;
        subfilter.temporalValue = 'after';
      } else {
        // find the key that field converts to
        subfilter.field = fieldValueObject[filterFieldMap[operator]] || fieldValueObject.in_string_array || '';
        // find the key that value converts to
        subfilter.value = fieldValueObject[filterValueMap[operator]] || fieldValueObject.search_string_array || '';
        if (typeof subfilter.value === 'object') {
          // if there is a literal value, we want to remove that layer
          subfilter.value = subfilter.value.literal;
        }
        if (typeof subfilter.value === 'boolean') {
          // if type is boolean, need to convert to string
          subfilter.value = subfilter.value
            ? 'True'
            : 'False';
        }
        if (subfilter.field) {
          // pull temporal value from field [resourceName, temporalValue, fieldName]
          const splitField = subfilter.field.split('.');
          if (ignoreTemporals) subfilter.temporalValue = 'after';
          else [, subfilter.temporalValue, subfilter.field] = splitField;
        }
      }
      convertedFilter.subfilters.push(subfilter);
    }
  });
  return { convertedFilter, uniqueIndex };
}

/**
 * Locates group within path to add or update
 *
 * @property {Object} filter - Current layer of query filter
 * @property {String} paths - Path of current filter in relation to entire query filter
 * @property {number} pathIndex - How far into group we have searched
 * @property {Boolean} depth - How deep the query filter is
 *
 * @returns {Object} Group within query filter that path points to
 */
export function findGroup(filter, paths, pathIndex, depth) {
  if (depth === 0) return filter;
  if (pathIndex === 0) return findGroup(filter, paths, pathIndex + 1, depth);

  const subfilter = filter.subfilters[paths[pathIndex]];

  if (pathIndex === depth) return subfilter;
  return findGroup(subfilter, paths, pathIndex + 1, depth);
}

/**
 * Determines if current filter string has a depth greater than our max allowed
 * depth for the basic editor. We also ignore 'not' parenthesis.
 * NOTE: Need to use basic for loop, as forEach is not available for strings
 *
 * @property {string} queryFilterString - stringified query filter
 * @property {Number} maxDepth - Maximum depth allowed in filter string
 *
 * @returns {Boolean} Whether current query filter is three layers of depth or less
 */
export function checkIfWithinThreeLayers(queryFilterString, maxDepth) {
  const testString = queryFilterString.replace(/\\"/g, '').replace(/"[^"]*"/g, '');
  let depth = 0;
  let ignoreNots = 0;
  let withinThreeLayers = true;
  for (let index = 0; index < testString.length; index += 1) {
    if (testString[index] === '(') {
      if (index > 0 && testString.charAt(index - 1) === '!') {
        ignoreNots += 1;
      } else {
        depth += 1;
        if (depth > maxDepth) {
          withinThreeLayers = false;
        }
      }
    } else if (testString.charAt(index) === ')') {
      if (ignoreNots) {
        ignoreNots -= 1;
      } else {
        depth -= 1;
      }
    }
  }
  return withinThreeLayers;
}

/**
 * Checks if IGA filter has depth greater than basic editor allows
 * @param {Object} filter Current IGA filter
 * @param {Number} currentDepth recursive integer uses to determine depth
 * @param {Number} maxDepth Maximum allowed depth
 * @returns Boolean of whether filter is with specified depth
 */
export function checkIGAFilterWithinNLayers(filter, currentDepth, maxDepth) {
  if (currentDepth > maxDepth) {
    return false;
  }
  const layer = Object.entries(filter)[0][1];
  return layer.every((rule) => {
    if (rule.or || rule.and) {
      return checkIGAFilterWithinNLayers(rule, currentDepth + 1, maxDepth);
    }
    return true;
  });
}

/**
 * Checks if stringified IGA filter contains key signifying it contains array value
 * @param {String} filter stringified filter to check if it includes array key
 * @returns Boolean of whether filter contains array key
 */
export function checkIGAFilterForArrays(filter) {
  return filter.includes('search_string_array');
}

/**
 * Checks if IGA filter contains matching before/after conditions for each property in same level
 * @param {String} filter IGA filter to validate
 * @returns Boolean of whether filter is valid
 */
export function checkIGAFilterMeetsRequirements(filter) {
  const layer = Object.values(filter)[0];
  const properties = {};
  layer.forEach((subfilter) => {
    // don't need to check for second condition if subfilter is an or or an and, or if equals or not_equals and both sides of comparison are fields
    if (!(
      subfilter.or
      || subfilter.and
      || (
        subfilter.equals && (
          (
            typeof subfilter.equals.left === 'string'
            && (subfilter.equals.left.includes('before') || subfilter.equals.left.includes('after'))
          )
          && (
            typeof subfilter.equals.right === 'string'
            && (subfilter.equals.right.includes('before') || subfilter.equals.right.includes('after'))
          )
        )
      )
      || (
        subfilter.not_equals && (
          (
            typeof subfilter.not_equals.left === 'string'
            && (subfilter.not_equals.left.includes('before') || subfilter.not_equals.left.includes('after'))
          )
          && (
            typeof subfilter.not_equals.right === 'string'
            && (subfilter.not_equals.right.includes('before') || subfilter.not_equals.right.includes('after'))
          )
        )
      )
    )) {
      // setup to determine if there is at least one after condition to match each unique before property
      const subfilterValues = Object.values(subfilter)[0];
      const value = (subfilterValues.value || subfilterValues.left || subfilterValues.in_string || subfilterValues.in_string_array);
      const propertyName = value.split('.')[2];
      if (!properties[propertyName]) {
        properties[propertyName] = {};
      }
      if (value.includes('before')) {
        properties[propertyName].beforeFound = true;
      } else if (value.includes('after')) {
        properties[propertyName].afterFound = true;
      }
    }
  });
  const allSubfiltersValid = layer.every((subfilter) => {
    if (subfilter.or || subfilter.and) {
      return checkIGAFilterMeetsRequirements(subfilter);
    }
    return true;
  });
  const currentLayerValid = Object.values(properties).every((property) => property.beforeFound && property.afterFound);
  return allSubfiltersValid && currentLayerValid;
}

/**
 * Checks if a key is a glossary attribute
 * A glossary attribute is a key that starts with catalog.{resource}.glossary.,
 * where {resource} is either application, entitlement or role.
 * @param {String} key The key to check
 * @returns {Boolean} Whether the key is a glossary attribute
 */
export function isGlossaryAttribute(key) {
  const pattern = /^catalog\.(application|entitlement|role)\.glossary\./;
  return pattern.test(key);
}
