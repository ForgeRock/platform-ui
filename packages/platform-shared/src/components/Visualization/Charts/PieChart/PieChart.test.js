/**
 * Copyright (c) 2022-2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import { shallowMount } from '@vue/test-utils';
import * as d3 from 'd3';
import PieChart from './index';

const mockData = [
  {
    label: 'item1',
    value: 30,
    color: 'red',
  },
  {
    label: 'item2',
    value: 20,
    color: 'yellow',
  },
  {
    label: 'item3',
    value: 20,
    color: 'yellow',
  },
];

const mountComponent = (props) => shallowMount(PieChart, {
  mocks: { $t: () => {} },
  propsData: {
    data: mockData,
    ...props,
  },
});

describe('PieChart', () => {
  let wrapper;

  it('PieChart successfully loaded', () => {
    wrapper = mountComponent();
    expect(wrapper.name()).toEqual('PieChart');
  });

  it('renders the chart div', () => {
    wrapper = mountComponent();
    const chart = findByTestId(wrapper, 'chart');
    expect(chart.exists()).toBe(true);
  });

  it('calls to create chart based on data', () => {
    wrapper = mountComponent();
    const createChartSpy = jest.spyOn(wrapper.vm, 'createChart');
    wrapper.vm.loadData();
    expect(createChartSpy).toHaveBeenCalledWith({
      0: 30,
      1: 20,
      2: 20,
    }, ['red', 'yellow', 'yellow']);
  });

  describe('legend', () => {
    it('renders a legend', () => {
      wrapper = mountComponent();
      const legend = findByTestId(wrapper, 'legend');
      expect(legend.exists()).toBe(true);
    });

    it('contains list elements for each data object', () => {
      wrapper = mountComponent();
      const legend = findByTestId(wrapper, 'legend');
      const items = legend.findAll('li');
      expect(items.length).toBe(3);
    });

    it('has correct label text for legend items', () => {
      wrapper = mountComponent();
      const legend = findByTestId(wrapper, 'legend');
      const items = legend.findAll('li');
      items.wrappers.forEach((item, index) => {
        expect(item.text()).toBe(mockData[index].label);
      });
    });

    it('has correct colors for legend items', () => {
      wrapper = mountComponent();
      const legend = findByTestId(wrapper, 'legend');
      const items = legend.findAll('span');
      items.wrappers.forEach((item, index) => {
        const bullet = item.find('.rounded-pill');
        expect(bullet.attributes('style')).toBe(`height: 10px; width: 24px; background-color: ${mockData[index].color};`);
      });
    });

    it('displays value in legend when showLegenCount is true', () => {
      wrapper = mountComponent({ showLegendCount: true });
      const legend = findByTestId(wrapper, 'legend');
      const items = legend.findAll('li');
      items.wrappers.forEach((item, index) => {
        expect(item.text()).toBe(`${mockData[index].label} (${mockData[index].value})`);
      });
    });

    it('applies custom class to legend based on legendClass prop', () => {
      wrapper = mountComponent({ legendClass: 'testClass' });
      const legend = findByTestId(wrapper, 'legend');
      expect(legend.attributes('class')).toBe('testClass');
    });
  });

  describe('tooltip', () => {
    let tooltip;

    beforeEach(() => {
      tooltip = {};
      tooltip.getBoundingClientRect = jest.fn().mockReturnValue({ width: 10, height: 10 });
      tooltip.node = jest.fn().mockReturnValue(tooltip);
      tooltip.style = jest.fn().mockReturnValue(tooltip);
    });

    it('selects the tooltip', () => {
      const selectAllSpy = jest.spyOn(d3, 'selectAll');
      wrapper = mountComponent({ id: 'test' });
      expect(selectAllSpy).toHaveBeenCalledWith('.test-tooltip');
    });

    it('shows tooltip for first data object', () => {
      wrapper = mountComponent();
      let html = '';
      tooltip.html = jest.fn().mockImplementation((cb) => {
        html = cb();
        return tooltip;
      });

      wrapper.vm.tooltip = tooltip;

      wrapper.vm.mouseMove({
        clientX: 100,
        clientY: 100,
        srcElement: {
          __data__: { index: 0 },
        },
      });

      expect(html).toEqual(expect.stringContaining('item1'));
      expect(html).toEqual(expect.stringContaining('30'));
    });

    it('shows tooltip for second object', () => {
      wrapper = mountComponent();
      let html = '';
      tooltip.html = jest.fn().mockImplementation((cb) => {
        html = cb();
        return tooltip;
      });

      wrapper.vm.tooltip = tooltip;

      wrapper.vm.mouseMove({
        clientX: 100,
        clientY: 100,
        srcElement: {
          __data__: { index: 1 },
        },
      });
      expect(html).toEqual(expect.stringContaining('item2'));
      expect(html).toEqual(expect.stringContaining('20'));
    });

    it('shows tooltip for third data object', () => {
      wrapper = mountComponent();
      let html = '';
      tooltip.html = jest.fn().mockImplementation((cb) => {
        html = cb();
        return tooltip;
      });

      wrapper.vm.tooltip = tooltip;

      wrapper.vm.mouseMove({
        clientX: 100,
        clientY: 100,
        srcElement: {
          __data__: { index: 2 },
        },
      });
      expect(html).toEqual(expect.stringContaining('item3'));
      expect(html).toEqual(expect.stringContaining('20'));
    });
  });
});
