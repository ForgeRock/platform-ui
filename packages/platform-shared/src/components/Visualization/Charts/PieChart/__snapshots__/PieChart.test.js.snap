// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart renders the chart div: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) "watch" option or vm.$watch on an array value will no longer trigger on array mutation unless the "deep" option is specified. If current usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ WATCH_ARRAY: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/watch.html",
    "
",
    " at <PieChart",
    "data=",
    [
      {
        "color": "red",
        "label": "item1",
        "value": 30,
      },
      {
        "color": "yellow",
        "label": "item2",
        "value": 20,
      },
      {
        "color": "yellow",
        "label": "item3",
        "value": 20,
      },
    ],
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
